src/core/net/dns_headers.cpp

```diff
@@ -293,6 +293,123 @@ otError Name::ReadName(const Message &aMessage,
     return error;
 }
 
+otError Name::CompareLabel(const Message &aMessage, uint16_t &aOffset, uint16_t aHeaderOffset, const char *aLabel)
+{
+    otError       error;
+    LabelIterator iterator(aMessage, aOffset, aHeaderOffset);
+
+    SuccessOrExit(error = iterator.GetNextLabel());
+    VerifyOrExit(iterator.CompareLabel(aLabel, /* aIsSingleLabel */ true), error = OT_ERROR_NOT_FOUND);
+    aOffset = iterator.mNextLabelOffset;
+
+exit:
+    return error;
+}
+
+otError Name::CompareName(const Message &aMessage, uint16_t &aOffset, uint16_t aHeaderOffset, const char *aName)
+{
+    otError       error;
+    LabelIterator iterator(aMessage, aOffset, aHeaderOffset);
+    bool          matches = true;
+
+    if (*aName == kLabelSeperatorChar)
+    {
+        aName++;
+        VerifyOrExit(*aName == kNullChar, error = OT_ERROR_INVALID_ARGS);
+    }
+
+    while (true)
+    {
+        error = iterator.GetNextLabel();
+
+        switch (error)
+        {
+        case OT_ERROR_NONE:
+            if (matches && !iterator.CompareLabel(aName, /* aIsSingleLabel */ false))
+            {
+                matches = false;
+            }
+
+            break;
+
+        case OT_ERROR_NOT_FOUND:
+            // We reached the end of the name in `aMessage`. We check if
+            // all the previous labels matched so far, and we are also
+            // at the end of `aName` string (see null char), then we
+            // return `OT_ERROR_NONE` indicating a successful comparison
+            // (full match). Otherwise we return `OT_ERROR_NOT_FOUND` to
+            // indicate failed comparison.
+
+            if (matches && (*aName == kNullChar))
+            {
+                error = OT_ERROR_NONE;
+            }
+
+            aOffset = iterator.mNameEndOffset;
+
+            OT_FALL_THROUGH;
+
+        default:
+            ExitNow();
+        }
+    }
+
+exit:
+    return error;
+}
+
+otError Name::CompareName(const Message &aMessage,
+                          uint16_t &     aOffset,
+                          uint16_t       aHeaderOffset,
+                          const Message &aMessage2,
+                          uint16_t       aOffset2,
+                          uint16_t       aHeaderOffset2)
+{
+    otError       error;
+    LabelIterator iterator(aMessage, aOffset, aHeaderOffset);
+    LabelIterator iterator2(aMessage2, aOffset2, aHeaderOffset2);
+    bool          matches = true;
+
+    while (true)
+    {
+        error = iterator.GetNextLabel();
+
+        switch (error)
+        {
+        case OT_ERROR_NONE:
+            // If all the previous labels matched so far, then verify
+            // that we can get the next label on `iterator2` and that it
+            // matches the label from `iterator`.
+            if (matches && (iterator2.GetNextLabel() != OT_ERROR_NONE || !iterator.CompareLabel(iterator2)))
+            {
+                matches = false;
```

Comment:
should we exit earlier when not matches?

Reply:
We cannot. We need to read/parse to end to be able to correctly update `aOffset` .

Note that `aOffset` is updated to skip over the name field in message (whether the read name does match the given name or not). This behavior (of skipping over the name when being compared) is actually very helpful when searching over a bunch of records for the one matching a name.